### ISSUE TITLE
'elst' box and timestamp handling in gapless playback

### ISSUE THREAD

**Issue by ybai001** [google/ExoPlayer#10503]
The test file is https://www2.iis.fraunhofer.de/AAC/gapless-sweep_part1_iis.m4a (Download it. The downloaded file suffix is .mp3 and you need to  rename the suffix to .m4a manually.) This file includes one 'elst' box with media_time = 1600. Timescale of this AAC track is 44100.

If I use ExoPlayer to play this media file, I'll get input timestamp from Google AAC software decoder (C2SoftAacDec) as below:
03-24 23:00:12.533  1251  4699 E C2SoftAacDec: [BYN3]inInfo.timestamp = 0    // CSD, just ignore it
03-24 23:00:12.536  1251  4699 E C2SoftAacDec: [BYN3]inInfo.timestamp = 1000000000000    // First sample timestamp, it is "0". The MSB "1" is added by ExoPlayer. We can ignore it at first. So the first sample timestamp is "0".
03-24 23:00:12.538  1251  4699 E C2SoftAacDec: [BYN3]inInfo.timestamp = 1000000023219    // Second sample timestamp, it is "23219". It comes from 1024 (frame size)/44100 * 1000000 = 23219
03-24 23:00:12.539  1251  4699 E C2SoftAacDec: [BYN3]inInfo.timestamp = 1000000046439

But If I use Android system media player to play this media file, I'll get input timestamp as below:

03-24 22:40:47.303  1265  5128 E C2SoftAacDec: [BYN3]inInfo.timestamp = 0    // CSD, just ignore it
03-24 22:40:47.303  1265  5128 E C2SoftAacDec: [BYN3]inInfo.timestamp = -36281    // First sample timestamp, it is "-36281". It comes from -1600 ('elst' box media_time value) / 44100 * 1000000 = -36281
03-24 22:40:47.304  1265  5128 E C2SoftAacDec: [BYN3]inInfo.timestamp = -13061
03-24 22:40:47.305  1265  5128 E C2SoftAacDec: [BYN3]inInfo.timestamp = 10158
03-24 22:40:47.307  1265  5128 E C2SoftAacDec: [BYN3]inInfo.timestamp = 33378

If I use ffprobe to get this mpeg-4 file info, I'll get

[PACKET]
codec_type=audio
stream_index=0
pts=-1600
pts_time=-0.036281
dts=-1600
dts_time=-0.036281
duration=1024
duration_time=0.023220
size=278
pos=1760
flags=KD
[SIDE_DATA]
side_data_type=Skip Samples
skip_samples=1600
discard_padding=0
skip_reason=0
discard_reason=0
[/SIDE_DATA]
[/PACKET]
[PACKET]
codec_type=audio
stream_index=0
pts=-576
pts_time=-0.013061
dts=-576
dts_time=-0.013061
duration=1024
duration_time=0.023220
size=336
pos=2038
flags=K_
[/PACKET]

In summary, considering the sample timestamp, AOSP and ffmpeg considers the 'elst' box value while ExoPlayer doesn't do it. I'd like to know which behavior is correct behavior. I went through the MPEG-4 spec but can't find answer so that I ask this question here. Thanks a lot in advance.

**Comment by ybai001** [comment]
BTW, the ffprobe command I used is `ffprobe.exe -i gapless-sweep_part1_iis.m4a -show_packets > out.txt`

**Comment by tonihei** [comment]
The `elst` box describes an edit list, specifying which parts of the file should be played.

Specific edit lists that only clip a short duration from the start or end of the file are considered as gapless metadata by ExoPlayer and are applied at a later stage. See [this code](https://github.com/google/ExoPlayer/blob/ab4d37f499ba49bb3cac7938eb03ebf7133ccfe6/library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/AtomParsers.java#L606) for details. So everything is working as intended as far as I can see. 

MediaPlayer doesn't this type of gapless metadata and just applies the edit list. Note that it produces negative sample timestamps in the process, which isn't well supported by many decoders and may cause playback issues.

**Comment by ybai001** [comment]
Thanks for quick feedback. Yes, I see ExoPlayer handles 'elst' box well.

I think Android AOSP also handles gapless metadata, see this code: https://cs.android.com/android/platform/superproject/+/master:frameworks/av/media/libstagefright/SkipCutBuffer.cpp;drc=0766ead67694809c3cb12dc6ddb81a2eb7b475ad;l=59
As long as AOSP MPEG4Extractor finds 'elst' box, one SkipCutBuffer will be created. It will be called for each codec output buffer. It trims the first "encoder-latency" samples (calculated based on elst media_time) and always keep "encoder-padding" samples in it.

I understood negative sample timestamps are not well supported by some decoders. In fact, the trick point here is Android AOSP itself generates negative timestamp for this mp4 file while AOSP AAC decoder input timestamp data type is "unsigned long". Fortunately, I have not identified any serious issue for this logic.

According to your reply, it seems that there is no standard to define whether 'elst' box info should be a factor of input timestamp. I know either of them are OK for most cases (Means absolute input timestamp value is NOT important). But for our specific use case, it is important. It seems that we need to handle both cases (first input frame timestamp could be 0 or -1600 for this file.

**Comment by tonihei** [comment]
I'll close this issue because I assume the question has been answered and you found a path forward.
